firebase/firestore.rules

```diff
@@ -1,8 +1,164 @@
+
+function isAuthenticated() {
+  return request.auth.uid != null;
+}
+
+function isUser(id) {
+  return isAuthenticated() && request.auth.uid == id;
+}
+
+function isVerifiedCAV() {
+  // TODO: Actually use auth claims
+  return isAuthenticated() && true;
+}
+
+function isVerifiedPIN() {
+  // TODO: Actually use auth claims
+  return isAuthenticated() && true;
+}
+
+function currentUserIsCurrentPINRef(userID, database, data) {
+  return isAuthenticated() &&
+  ('pinUserRef' in data) &&
+  data.pinUserRef == /databases/$(database)/documents/users/$(userID)
+}
+
+function currentUserIsCurrentCAVRef(userID, database, data) {
+  return isAuthenticated() &&
+  ('cavUserRef' in data) &&
+  data.cavUserRef == /databases/$(database)/documents/users/$(userID)
+}
+
+function questionnaireBelongsToUser(userID, database, data) {
+    return isAuthenticated() &&
+           ('parentRef' in data) &&
+           data.parentRef == /databases/$(database)/documents/users/$(userID)
+}
+
+function confirmEitherPINorCAV(userID, database, data) {
+    return currentUserIsCurrentPINRef(userID, database, data) || currentUserIsCurrentCAVRef(userID, database, data)
+}
+
+function isNotSameCAVPIN(data) {
+  return data.cavUserRef != data.pinUserRef
+}
+
+function verifyUserDataComplete(data) {
+   return data.averageRating >= 1 &&
+          data.casesCompleted >= 0 &&
+          data.requestsMade >= 0 &&
+          ('username' in data)
+}
+
+function verifyQuestionnaireComplete(data) {
+   return ('parentRef' in data) &&
+          ('data' in data) &&
+          ('type' in data) &&
+          ('createdAt' in data) &&
+          ('version' in data)
+}
+
+// Completed requests need ratings on all sides
+function verifyHelpRequestAttributesForStatus(status, data) {
+    // No if/else so we have to do this
+    return (
+             status == 'ongoing' && ('cavUserRef' in data)
+           ) ||
+           (
+             (status == 'completed') &&
+             ('cavUserRef' in data) &&
+             ('pinRating' in data) &&
+             ('cavRating' in data) &&
+             ('pinRatedAt' in data) &&
+             ('cavRatedAt' in data)
+           ) ||
+           true // for other statuses we don't have extra requirements
+}
+
+function verifyHelpRequestDataComplete(data) {
+   return ('pinUserRef' in data) &&
+          ('pinUserSnapshot' in data) &&
+          ('title' in data) &&
+          ('description' in data) &&
+          ('latLng' in data) &&
+          ('status' in data) &&
+          ('createdAt' in data) &&
+          ('updatedAt' in data) &&
+          verifyHelpRequestAttributesForStatus(data.status, data)
+}
+
+function verifyOfferDataComplete(data) {
+   return ('cavUserRef' in data) &&
+          ('pinUserRef' in data) &&
+          ('requestRef' in data) &&
+          ('cavUserSnapshot' in data) &&
+          ('message' in data) &&
+          ('status' in data) &&
+          ('createdAt' in data)
+}
+
+// TODO: Further validate address object.
```

Comment:
Won't this kind of validation be hard to implement only with boolean expressions? Shouldn't this be performed in a cloud function?

Reply:
yea, but we have to do some kind of basic validation here maybe just to protect the db since clients can write to it directly.